src/Agent.Sdk/SecretMasking/OssSecretMasker.cs

```diff
@@ -81,11 +106,132 @@ public void Dispose()
+
+        if (events < MaxTelemetryDetectionEvents && detectionData != null)
+        {
+            publishAction("SecretMaskerDetections", detectionData);
+            detectionData = null;
+        }
+
+        var overallData = new Dictionary<string, string>(GetOverallTelemetry());
+        if (_telemetry.Detections.Count > MaxTelemetryDetections)
+        {
+            overallData.Add("DetectionDataIsIncomplete", "true");
+        }
+
+        publishAction("SecretMasker", overallData);
+    }
+
+    private KeyValuePair<string, string>[] GetOverallTelemetry()
+    {
+        double elapsedMaskingTimeInMilliseconds = (double)_secretMasker.ElapsedMaskingTime / TimeSpan.TicksPerMillisecond;
+        return new KeyValuePair<string, string>[] {
+            new("Version", SecretMasker.Version.ToString()),
+            new("CharsScanned", _telemetry.CharsScanned.ToString(CultureInfo.InvariantCulture)),
+            new("StringsScanned", _telemetry.StringsScanned.ToString(CultureInfo.InvariantCulture)),
+            new("TotalDetections", _telemetry.TotalDetections.ToString(CultureInfo.InvariantCulture)),
+            new("UniqueCorrelatingIds", _telemetry.Detections.Count.ToString(CultureInfo.InvariantCulture)),
+            new("ElapsedMaskingTimeInMilliseconds", elapsedMaskingTimeInMilliseconds.ToString(CultureInfo.InvariantCulture)),
+        };
+    }
+
+    private sealed class Telemetry
+    {
+        private readonly ConcurrentDictionary<string, string> _detections;
+        private long _charsScanned;
+        private long _stringsScanned;
+        private long _totalDetections;
+
+        public IReadOnlyDictionary<string, string> Detections => _detections;
+        public long CharsScanned => _charsScanned;
+        public long StringsScanned => _stringsScanned;
+        public long TotalDetections => _totalDetections;
+        public readonly Action<Detection> DetectionAction;
+
+        public Telemetry()
+        {
+            _detections = new ConcurrentDictionary<string, string>();
+            DetectionAction = ProcessDetection;
+        }
+
+        private Telemetry(Telemetry copy)
+        {
+            _detections = new ConcurrentDictionary<string, string>(copy.Detections);
+            DetectionAction = ProcessDetection;
+
+            _charsScanned = copy.CharsScanned;
+            _stringsScanned = copy.StringsScanned;
+            _totalDetections = copy.TotalDetections;
+        }
+
+        public Telemetry Clone() => new(this);
+
+        public void ProcessInput(string input)
+        {
+            Interlocked.Add(ref _charsScanned, input.Length);
+            Interlocked.Increment(ref _stringsScanned);
+        }
+
+        private void ProcessDetection(Detection detection)
+        {
+            Interlocked.Increment(ref _totalDetections);
+
+            if (detection.CrossCompanyCorrelatingId != null)
+            {
+                _detections.TryAdd(detection.CrossCompanyCorrelatingId, detection.Moniker);
```

Comment:
>TryAdd [](http://example.com/codeflow?start=28&length=6)

I note here that you will continue to collect novel observed c3ids even after we've exceeded the configured capacity to raise events. With another design you'd stop collecting them (though you could still collect the other metrics, raw detections count, string count, chars count). And then when you publish them, you'd simply process all detections with no need to track when you exceeded the max limit.

Reply:
We now stop collecting when threshold is exceeded. This removes a small feature that reported how many unique C3IDs were observed irrespective including in excess of the cap. But I'm not sure what we would do with this information and it does seem prudent to constrain the memory usage of the feature as we do now.